### PR TITLE
gh-101006: Improve error handling when read marshal data

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
* EOFError no longer overrides other errors such as MemoryError or OSError at the start of the object.
* Raise more relevant error when the NULL object occurs as a code object component.
* Minimize an overhead of calling PyErr_Occurred().

* Issue: gh-101006

### CODE REVIEW

File: Python/marshal.c [hunk 1]
```diff
@@ -824,10 +829,10 @@ r_PyLong(RFILE *p)
     digit d;
 
     n = r_long(p);
-    if (PyErr_Occurred())
-        return NULL;
     if (n == 0)
         return (PyObject *)_PyLong_New(0);
+    if (n == -1 && PyErr_Occurred())
+        return NULL;
```
Comment (erlend-aasland) [inline Python/marshal.c]:
```suggestion
    if (n == -1 && PyErr_Occurred()) {
        return NULL;
    }
```

File: Python/marshal.c [hunk 2]
```diff
@@ -875,18 +872,16 @@ r_PyLong(RFILE *p)
         }
         d += (digit)md << j*PyLong_MARSHAL_SHIFT;
     }
-    if (PyErr_Occurred()) {
-        Py_DECREF(ob);
-        return NULL;
-    }
```
Comment (erlend-aasland) [inline Python/marshal.c:881]: Perhaps leave an assert:

```suggestion
    assert(!PyErr_Occurred());
```

File: Python/marshal.c [hunk 3]
```diff
@@ -1036,7 +1031,9 @@ r_object(RFILE *p)
 
     case TYPE_INT:
         n = r_long(p);
-        retval = PyErr_Occurred() ? NULL : PyLong_FromLong(n);
+        if (n == -1 && PyErr_Occurred())
+            break;
```
Comment (erlend-aasland) [inline Python/marshal.c]:
```suggestion
        if (n == -1 && PyErr_Occurred()) {
            break;
        }
```

File: Python/marshal.c [hunk 4]
```diff
@@ -1102,10 +1099,10 @@ r_object(RFILE *p)
         {
             const char *ptr;
             n = r_long(p);
-            if (PyErr_Occurred())
-                break;
             if (n < 0 || n > SIZE32_MAX) {
-                PyErr_SetString(PyExc_ValueError, "bad marshal data (bytes object size out of range)");
+                if (!PyErr_Occurred()) {
+                    PyErr_SetString(PyExc_ValueError, "bad marshal data (bytes object size out of range)");
```
Comment (erlend-aasland) [inline Python/marshal.c]:
```suggestion
                    PyErr_SetString(PyExc_ValueError,
                        "bad marshal data (bytes object size out of range)");
```

File: Python/marshal.c [hunk 5]
```diff
@@ -1127,10 +1124,10 @@ r_object(RFILE *p)
         /* fall through */
     case TYPE_ASCII:
         n = r_long(p);
-        if (PyErr_Occurred())
-            break;
         if (n < 0 || n > SIZE32_MAX) {
-            PyErr_SetString(PyExc_ValueError, "bad marshal data (string size out of range)");
+            if (!PyErr_Occurred()) {
+                PyErr_SetString(PyExc_ValueError, "bad marshal data (string size out of range)");
```
Comment (erlend-aasland) [inline Python/marshal.c]:
```suggestion
                PyErr_SetString(PyExc_ValueError,
                    "bad marshal data (string size out of range)");
```

File: Python/marshal.c [hunk 6]
```diff
@@ -1169,10 +1164,10 @@ r_object(RFILE *p)
         const char *buffer;
 
         n = r_long(p);
-        if (PyErr_Occurred())
-            break;
         if (n < 0 || n > SIZE32_MAX) {
-            PyErr_SetString(PyExc_ValueError, "bad marshal data (string size out of range)");
+            if (!PyErr_Occurred()) {
+                PyErr_SetString(PyExc_ValueError, "bad marshal data (string size out of range)");
```
Comment (erlend-aasland) [inline Python/marshal.c]:
```suggestion
                PyErr_SetString(PyExc_ValueError,
                    "bad marshal data (string size out of range)");
```

File: Python/marshal.c [hunk 7]
```diff
@@ -1194,16 +1189,17 @@ r_object(RFILE *p)
...
+        }
         goto _read_tuple;
     case TYPE_TUPLE:
         n = r_long(p);
-        if (PyErr_Occurred())
-            break;
         if (n < 0 || n > SIZE32_MAX) {
-            PyErr_SetString(PyExc_ValueError, "bad marshal data (tuple size out of range)");
+            if (!PyErr_Occurred()) {
+                PyErr_SetString(PyExc_ValueError, "bad marshal data (tuple size out of range)");
```
Comment (erlend-aasland) [inline Python/marshal.c]:
```suggestion
                PyErr_SetString(PyExc_ValueError,
                    "bad marshal data (tuple size out of range)");
```

File: Python/marshal.c [hunk 8]
```diff
@@ -1228,10 +1224,10 @@ r_object(RFILE *p)
 
     case TYPE_LIST:
         n = r_long(p);
-        if (PyErr_Occurred())
-            break;
         if (n < 0 || n > SIZE32_MAX) {
-            PyErr_SetString(PyExc_ValueError, "bad marshal data (list size out of range)");
+            if (!PyErr_Occurred()) {
+                PyErr_SetString(PyExc_ValueError, "bad marshal data (list size out of range)");
```
Comment (erlend-aasland) [inline Python/marshal.c]:
```suggestion
                PyErr_SetString(PyExc_ValueError,
                    "bad marshal data (list size out of range)");
```

File: Python/marshal.c [hunk 9]
```diff
@@ -1284,10 +1280,10 @@ r_object(RFILE *p)
     case TYPE_SET:
     case TYPE_FROZENSET:
         n = r_long(p);
-        if (PyErr_Occurred())
-            break;
         if (n < 0 || n > SIZE32_MAX) {
-            PyErr_SetString(PyExc_ValueError, "bad marshal data (set size out of range)");
+            if (!PyErr_Occurred()) {
+                PyErr_SetString(PyExc_ValueError, "bad marshal data (set size out of range)");
```
Comment (erlend-aasland) [inline Python/marshal.c]:
```suggestion
                PyErr_SetString(PyExc_ValueError,
                    "bad marshal data (set size out of range)");
```

File: Python/marshal.c [hunk 10]
```diff
@@ -1467,9 +1467,9 @@ r_object(RFILE *p)
     case TYPE_REF:
         n = r_long(p);
         if (n < 0 || n >= PyList_GET_SIZE(p->refs)) {
-            if (n == -1 && PyErr_Occurred())
-                break;
-            PyErr_SetString(PyExc_ValueError, "bad marshal data (invalid reference)");
+            if (!PyErr_Occurred()) {
+                PyErr_SetString(PyExc_ValueError, "bad marshal data (invalid reference)");
```
Comment (erlend-aasland) [inline Python/marshal.c]:
```suggestion
                PyErr_SetString(PyExc_ValueError,
                    "bad marshal data (invalid reference)");
```

**Review by erlend-aasland — COMMENTED** [review]
After a quick review, this looks good to me. I'd like to take a closer look if you can wait for a day or two.

File: Python/marshal.c (hunk 2 above)
Comment (serhiy-storchaka) [reply Python/marshal.c:881]: Applying suggestions on deleted lines is currently not supported. :-(

Comment (erlend-aasland) [reply Python/marshal.c:881]: That's strange! The "Commit suggestion" button seems to be enabled in my browser.

Comment (serhiy-storchaka) [reply Python/marshal.c:881]: Could you please commit it?

Comment (arhadthedev) [reply Python/marshal.c:881]: > That's strange! The "Commit suggestion" button seems to be enabled in my browser.

It allows to enter a commit message but fails on the next step:

> image

Comment (arhadthedev) [reply Python/marshal.c:881]: The editor opened from the PR files view (the ellipsis button → Edit file) seems to work better for such changes.

Comment (erlend-aasland) [reply Python/marshal.c:881]: > Could you please commit it?

It seems @arhadthedev already did so.

Comment (serhiy-storchaka) [reply Python/marshal.c:881]: Thanks, @arhadthedev.

**Pull request merged**